x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -159,10 +159,55 @@ private Predicate<IndexAbstraction> buildIndexMatcherPredicateForAction(String a
         }
         final StringMatcher nameMatcher = indexMatcher(ordinaryIndices, restrictedIndices);
         final StringMatcher bwcSpecialCaseMatcher = indexMatcher(grantMappingUpdatesOnIndices, grantMappingUpdatesOnRestrictedIndices);
-        return indexAbstraction -> nameMatcher.test(indexAbstraction.getName())
-            || (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM
-                && (indexAbstraction.getParentDataStream() == null)
-                && bwcSpecialCaseMatcher.test(indexAbstraction.getName()));
+        return new IsResourceAuthorizedPredicate(nameMatcher, bwcSpecialCaseMatcher);
+    }
+
+    /**
+     * This encapsulates the authorization test for resources.
+     * There is an additional test for resources that are missing or that are not a datastream or a backing index.
+     */
+    public static class IsResourceAuthorizedPredicate {
+
+        private final StringMatcher resourceNameMatcher;
+        private final StringMatcher additionalNonDatastreamNameMatcher;
+
+        // public for tests
+        public IsResourceAuthorizedPredicate(StringMatcher resourceNameMatcher, StringMatcher additionalNonDatastreamNameMatcher) {
+            this.resourceNameMatcher = resourceNameMatcher;
+            this.additionalNonDatastreamNameMatcher = additionalNonDatastreamNameMatcher;
+        }
+
+        /**
+         * Authorization test for resources.
+         * An additional predicate tests datastreams and backing indices, apart from the regular predicate that tests all resource types.
+         */
+        public final boolean test(IndexAbstraction indexAbstraction) {
```

Comment:
for public methods that return boolean values, can you include "returns true when X , false when Y (or false otherwise)"